src/raft/consensus_state_machine.nim

```diff
@@ -97,320 +93,529 @@ type
+  ps.term() == sm.term and ps.votedFor() == sm.votedFor and
+    ps.commitIndex() == sm.commitIndex and ps.persistedIndex() == sm.log.lastIndex
+
+func leader*(sm: RaftStateMachineRef): var LeaderState =
+  sm.state.leader
+
+func follower*(sm: RaftStateMachineRef): var FollowerState =
+  sm.state.follower
+
+func candidate*(sm: RaftStateMachineRef): var CandidateState =
+  sm.state.candidate
+
+func isLeader*(sm: RaftStateMachineRef): bool =
+  sm.state.isLeader
+
+func isFollower*(sm: RaftStateMachineRef): bool =
+  sm.state.isFollower
+
+func isCandidate*(sm: RaftStateMachineRef): bool =
+  sm.state.isCandidate
+
+const loglevel {.intdefine.}: int = int(DebugLogLevel.Error)
+
+template addDebugLogEntry(
+    smArg: RaftStateMachineRef, levelArg: DebugLogLevel, messageArg: string
+) =
+  let sm = smArg
+  let level = levelArg
+  let message = messageArg
+  if loglevel >= int(level):
+    sm.output.debugLogs.add(
+      DebugLogEntry(
+        time: sm.timeNow,
+        state: sm.state.state,
+        term: sm.term,
+        level: level,
+        msg: message,
+        nodeId: sm.myId,
+      )
+    )
+
+template debug*(sm: RaftStateMachineRef, log: string) =
   sm.addDebugLogEntry(DebugLogLevel.Debug, log)
 
-func warning*(sm: var RaftStateMachine, log: string) = 
+template warning*(sm: RaftStateMachineRef, log: string) =
   sm.addDebugLogEntry(DebugLogLevel.Warning, log)
 
-func error*(sm: var RaftStateMachine, log: string) = 
+template error*(sm: RaftStateMachineRef, log: string) =
   sm.addDebugLogEntry(DebugLogLevel.Error, log)
 
-func info*(sm: var RaftStateMachine, log: string) = 
+template info*(sm: RaftStateMachineRef, log: string) =
   sm.addDebugLogEntry(DebugLogLevel.Info, log)
 
-func critical*(sm: var RaftStateMachine, log: string) = 
+template critical*(sm: RaftStateMachineRef, log: string) =
   sm.addDebugLogEntry(DebugLogLevel.Critical, log)
 
-func resetElectionTimeout*(sm: var RaftStateMachine) =
-  # TODO actually pick random time
-  sm.randomizedElectionTime = sm.electionTimeout + times.initDuration(milliseconds = 100 + sm.randomGenerator.rand(200))
-
-func initRaftStateMachine*(id: RaftnodeId, currentTerm: RaftNodeTerm, log: RaftLog, commitIndex: RaftLogIndex, config: RaftConfig, now: times.DateTime, randomGenerator: Rand): RaftStateMachine = 
-  var sm =  RaftStateMachine()
+func getLeaderId*(sm: RaftStateMachineRef): Option[RaftNodeId] =
+  if sm.state.isLeader:
+    return some(sm.myId)
+  elif sm.state.isFollower:
+    return some(sm.state.follower.leader)
+  none(RaftNodeId)
+
+func observe*(ps: var RaftLastPollState, sm: RaftStateMachineRef) =
+  ps.setTerm sm.term
+  ps.setVotedFor sm.votedFor
+  ps.setCommitIndex sm.commitIndex
+  ps.setPersistedIndex sm.log.lastIndex
+
+func replicationStatus*(sm: RaftStateMachineRef): string =
+  var progressStrings = newSeqOfCap[string](sm.leader.tracker.progress.len)
+
+  let reportPreamble = "\nReplication report\n"
```

Comment:
Using `const` instead of `let` here will remove the run-time allocation for this string.
Ditto for `separator`

Reply:
done